R/Vietnam.R

```diff
@@ -0,0 +1,124 @@
+                         
+                         # Core Attributes (amend each paramater for country specific infomation)
+                         #' @field origin name of country to fetch data for
+                         origin = "Vietnam",
+                         #' @field supported_levels List of supported levels.
+                         supported_levels = list("1"),
+                         #' @field supported_region_names List of region names in order of level.
+                         supported_region_names = list("1" = "region"),
+                         #' @field supported_region_codes List of region codes in order of level.
+                         supported_region_codes = list("1" = "is_3166_2"),
+                         #' @field common_data_urls List of named links to raw data.
+                         # nolint start
+                         common_data_urls = list(
+                           "case_by_time" = 'https://covid.ncsc.gov.vn/api/v3/covid/provinces?filter_type=case_by_time',
+                           "death_by_time" = 'https://covid.ncsc.gov.vn/api/v3/covid/provinces?filter_type=death_by_time',
+                           "recovered_by_time" = 'https://covid.ncsc.gov.vn/api/v3/covid/provinces?filter_type=recovered_by_time'
+                           
+                         ),
+                         # nolint end
+                         #' @field source_data_cols existing columns within the raw data
+                         source_data_cols = c(
+                           "cases_total", "deaths_total", "recovered_total"
+                         ),
+                         #' @field source_text Plain text description of the source of the data
+                         source_text = "Public COVID-19 data curated by 5F team",
+                         #' @field source_url Website address for explanation/introduction of the
+                         #' data
+                         source_url = "https://covid.ncsc.gov.vn", # nolint
+                         
+                         #' @description Set up a table of region codes for clean data
+                         #' @importFrom tibble tibble
+                         set_region_codes = function(){
+                           self$codes_lookup$`1` <- covidregionaldata::vietnam_codes
+                         },
+                         
+                         #' @description Provincial Level Data
+                         #' cleaning
+                         #' @param ... pass additional arguments
+                         #'
+                         #' @importFrom dplyr filter select mutate rename
+                         #' @importFrom tidyr replace_na drop_na
+                         #' @importFrom lubridate dmy 
+                         #' @importFrom jsonlite fromJSON
+                         clean_common = function() {
+                           Sys.setenv("VROOM_CONNECTION_SIZE" = 131072*4) # Fix VROOM error 
```

Comment:
This looks interesting - what vroom error is this fixing? [We've had some other persistent problems with vroom on other datasets so we may want to put this into other classes or make it more general as behaviour.]

Reply:
It is to fix this kind of error:
```
2021-09-15 12:19:35 ERROR Error: The size of the connection buffer (131072) was not large enough
to fit a complete line:
  * Increase it by setting `Sys.setenv("VROOM_CONNECTION_SIZE")`
```
I will need to check one more time to see that is the correct place to put the line. This is also in connection with implementing the download part in the `download`.

Reply:
The wiki page gives some description of how the `data_url` list is used to slurp down csv files. 

https://github.com/epiforecasts/covidregionaldata/wiki/Adding-a-regional-data-source

If the data isn't a csv file then I think the list can safely be left blank, particularly if you override the `download` function. I am guessing that this `vroom` error may be coming from trying to apply it to a json file which doesn't have line breaks.

Reply:
Yes think Richard is right on this one and agree the solution is to overload the download method with one that can handle JSON. Really like your idea from #409 to have an alternative download option to handle JSON. Not sure what this would look like but could be as simple as having a dispatch in download to lower level functions (i.e not in the S6 class) that either read in a memoised csv or JSON based on the file ending. Obviously keen to here other ideas.

Reply:
A possibility would be to use `httr::GET() %>% httr::content()`. [As detailed in the help](https://httr.r-lib.org/reference/content.html#details), it will automatically dispatch to various parsers (including `readr::read_csv()` or `jsonlite::toJSON()`) depending on the declared mime type.

This assumes that mime types will be declared correctly, which may or may not be a strong assumption.

Reply:
I think that data class implementers usually have a good sense of whether they’re loading CSV or JSON, though I suppose you could want to load some of each.

I think we could look at writing something generic like our csv reader function which could then work through a list of urls (as in this Vietnam case) and generate a list of tables in `self$data$raw`. My concern is that JSON loading isn’t quite as generic as slurping up a CSV and sometimes you might be dealing with a REST implementation (which may enforce paging) and sometimes you may just be grabbing a large JSON (and cannot do paging).

Today I think I/we/@biocyberman can try to write a generic JSON downloader which works for Vietnam and acts like the CSV downloader we have. Making it more flexible can be a next step, with the condition that it doesn’t break the Vietnam code.

Reply:
Sounds like a good solution for a MVP Richard though Hugo I had no idea content worked like that - how cool!